cluster-controller/src/main/java/io/strimzi/controller/cluster/resources/Labels.java

```diff
@@ -0,0 +1,222 @@
+        return new Labels(userLabels);
+    }
+
+    /**
+     * Returns the labels of the given {@code resource}.
+     */
+    public static Labels fromResource(HasMetadata resource) {
+        return new Labels(resource.getMetadata().getLabels());
+    }
+
+    private Labels(Map<String, String> labels) {
+        this.labels = unmodifiableMap(new HashMap(labels));
+    }
+
+    private Labels with(String label, String value) {
+        Map<String, String> newLabels = new HashMap<>(labels.size() + 1);
+        newLabels.putAll(labels);
+        newLabels.put(label, value);
+        return new Labels(newLabels);
+    }
+
+    private Labels without(String label) {
+        Map<String, String> newLabels = new HashMap<>(labels);
+        newLabels.remove(label);
+        return new Labels(newLabels);
+    }
+
+    /**
+     * The same labels as this instance, but with the given {@code type} for the {@code strimzi.io/type} key.
+     */
+    public Labels withType(String type) {
+        return with(STRIMZI_TYPE_LABEL, type);
+    }
+
+    /**
+     * The same labels as this instance, but without any {@code strimzi.io/type} key.
+     */
+    public Labels withoutType() {
+        return without(STRIMZI_TYPE_LABEL);
+    }
+
+    /**
+     * The same labels as this instance, but with the given {@code kind} for the {@code strimzi.io/kind} key.
+     */
+    public Labels withKind(String kind) {
+        return with(STRIMZI_KIND_LABEL, kind);
+    }
+
+    /**
+     * The same labels as this instance, but without any {@code strimzi.io/kind} key.
+     */
+    public Labels withoutKind() {
+        return without(STRIMZI_KIND_LABEL);
+    }
+
+    /**
+     * The same labels as this instance, but with the given {@code cluster} for the {@code strimzi.io/cluster} key.
+     */
+    public Labels withCluster(String cluster) {
+        return with(STRIMZI_CLUSTER_LABEL, cluster);
+    }
+
+    /**
+     * The same labels as this instance, but with the given {@code name} for the {@code strimzi.io/name} key.
+     */
+    public Labels withName(String name) {
+        return with(STRIMZI_NAME_LABEL, name);
+    }
+
+    /**
+     * An unmodifiable map of the labels.
+     */
+    public Map<String, String> toMap() {
+        return labels;
+    }
+
+    /**
+     * A singleton instance with the given {@code cluster} for the {@code strimzi.io/cluster} key.
+     */
+    public static Labels cluster(String cluster) {
+        return new Labels(singletonMap(STRIMZI_CLUSTER_LABEL, cluster));
+    }
+
+    /**
+     * A singleton instance with the given {@code type} for the {@code strimzi.io/type} key.
+     */
+    public static Labels type(String type) {
+        return new Labels(singletonMap(STRIMZI_TYPE_LABEL, type));
+    }
+
+    /**
+     * A singleton instance with the given {@code kind} for the {@code strimzi.io/kind} key.
+     */
+    public static Labels kind(String kind) {
+        return new Labels(singletonMap(STRIMZI_KIND_LABEL, kind));
+    }
```

Comment:
I guess that if we change `clusterLabel' to `cluster` and `cluster` to `forCluster` we should do the same for the above `type` and `kind` and related `typeLabel` and `kindLabel`.